### ISSUE TITLE
repeatAt

### ISSUE THREAD

**Issue by gabrieledarrigo** [agenda/agenda#99]
Hello guys;
if I set a job like this:

<pre>
var job = agenda.create(type);
job.repeatAt('at 19:43');
job.save();
</pre>

After 19:43 (when the job has done its work) if i see the document saved into MongoDB I take this:

<pre>
> db.jobs.find().pretty();
{
        "_id" : ObjectId("5432d5903b8004b017dfeff5"),
        "name" : "someTask",
        "data" : null,
        "type" : "normal",
        "priority" : 0,
        "nextRunAt" : null,
        "repeatAt" : "at 19:48",
        "lastModifiedBy" : null,
        "lockedAt" : null,
        "lastRunAt" : ISODate("2014-10-06T17:48:00.998Z"),
        "lastFinishedAt" : ISODate("2014-10-06T17:48:01Z"),
        "failReason" : "failed to calculate repeatAt time due to invalid format",
        "failedAt" : ISODate("2014-10-06T17:48:00.998Z")
}
</pre>

with a fail reason.
So I believe tha job will not reRun the tomorrow day at 19:43, right?
Any hint?

Gabri

**Comment by bestan** [comment]
@rschmukler @gabrieledarrigo 
Actually date.js module does not set milliseconds to 0 when parsing values. Therefore, if job is completed within 1 second (not 1ms), you get "failed to calculate repeatAt time due to invalid format".

Run following script to see that milliseconds always match:

```javascript
var date = require("date.js");
setInterval(function() {
	console.log(Date.now(), date('19:43').valueOf())
}, 100);
```

I have worked around this issue by adding "tomorrow", e.g. repeatAt("tomorrow 11:00am")

**Comment by nwkeeley** [comment]
Apologies just looking at this now.... I think there is still an issue with how computeFromRepeatAt() returns the this.attrs.nextRunAt... I am going to try and investigate this weekend.

**Comment by btknorr** [comment]
Any update on this issue?  I get this failure every single time with repeatAt:

"failed to calculate repeatAt time due to invalid format"

**Comment by glesage** [comment]
Hey all, I am also getting this error, any updates? Thanks!

**Comment by glesage** [comment]
Ok it actually seems like this is really Date.js's fault (https://github.com/datejs/Datejs): if the string passed to `date()` is bad, it will return the current date... even though really it should be failing... seems really silly.

Since agenda checks every minute from what I can gather, if you set `job.repeatAt('9:00pm'); then when it runs at 21:00:00 it will equal the current date, 21:00:00

The fix for this is to set seconds AFTER 0 to the job: `job.repeatAt('9:00:01pm');` so when it runs at 21:00:00 it will not be equal to 21:00:01

Can anybody confirm?

**Comment by btknorr** [comment]
Hey all I got around this issue by setting it up like so:

```
var job = agenda.create('sync');
job.schedule('today at 9:00pm');
job.repeatAt('tomorrow 9:00pm');
```

**Comment by glesage** [comment]
Hm nice @btknorr. Does it work tomorrow for sure?

**Comment by btknorr** [comment]
yep been using this for a while now :)

**Comment by glesage** [comment]
ok awesome thanks!

**Comment by glesage** [comment]
I got the same error btw...

`failed to calculate repeatAt time due to invalid format`

Turns out my method mentioned above is also not working

**Comment by avivroth** [comment]
Is there any case where you can't replace "job.repeatAt" with a clever use of "job.repeatEvery" and "job.schedule"? e.g. instead of job.repeatAt('1:00am') you could set job.schedule('tomorrow at 1:00am') and job.repeatEvery('1 day') and get the desired effect.  Am I missing something?

**Comment by glesage** [comment]
Yes I believe this will work. Personally though I try not to rely on job nesting.. to be safe (:

**Comment by nwkeeley** [comment]
@avivroth if you do your approach and the job runs at 1am it will then reschedule and run again at 1am the next day.

However if the server goes down at 12am, comes back online at 5am your job will run right away. The problem with repeatEvery is that it simply adds 24hrs to the last time it executed.

RepeatAt maintains the time even if there is a server outage.

**Comment by ka-sh** [comment]
I went through the same issue, however, @btknorr solution worked well for me.
I am not doing anything fancy with date formats, following the same example as showing in the repeatAt section in documentations

`agenda.define(name, function (job, done) {`
      `logger.info('Executing job with data: %j', job.attrs.data);`
      `setTimeout(function () {`
       `console.log('Processing done...');`
       ` done();`
     ` }, 10000);`
    `});`

`var job = agenda.create(name, data);`
`job.repeatAt('3:30pm');`
`job.save();`

As you can see I tried to put done in a callback to increase processing time as fast execution might be one of the reasons as pointed by @rschmukler , however, it didn't solve the issue.

Could that be related to  function(job,done) doesn't block until done gets called ?

>  Async Job:
> 
> agenda.define('some long running job', function(job, done) {
>   doSomelengthyTask(function(data) {
>     formatThatData(data);
>     sendThatData(data);
>     done();
>   });
> });
> 
> Sync Job:
> 
> agenda.define('say hello', function(job) {
>   console.log("Hello!");
> });

**Comment by OmgImAlexis** [comment]
@agenda/maintainers I think we need to add better validation for time/date formats so invalid formats are detected before adding to the db.

**Comment by sonnguyenvan** [comment]
Hi all. Did you fixed this problem ?

**Comment by rogueturnip** [comment]
No, not yet, I've gotten busy with some other things and haven't had a chance to get back to it.

**Comment by dmitryrusakov** [comment]
Worked for me after using cron time format instead of human-readable: `'2:30am'` -> `'30 2 * * *'`. This approach works fine with time zone as well.

**Comment by tiendq** [comment]
Surprisingly this weird issue has been opened for 4 years :O 

Thanks @btknorr

**Comment by tehsunnliu** [comment]
2021 now and I'm still getting this error when setting the time as mentioned in the document `job.repeatAt("3:30pm");`

**Comment by tehsunnliu** [comment]
> Hey all I got around this issue by setting it up like so:
> 
> ```
> var job = agenda.create('sync');
> job.schedule('today at 9:00pm');
> job.repeatAt('tomorrow 9:00pm');
> ```

Saving the job as shown in Case 1 saves the last instance only. In this case, 'tomorrow 10:30am' only. Then I tried saving as mentioned in Case 2, both the instances get saved however I get the same error "failed to calculate repeatAt time due to invalid format". I even tried setting `cron format`, but setting this makes the job keep on running like an infinite loop.

Case 1.
```
job.schedule('today at 10:30am'); 
job.repeatAt('tommorow at 10:30am'); 
job.save();
```

Case 2
```
job.schedule('today at 10:30am').save(); 
job.repeatAt('tommorow at 10:30am').save(); 
```

Case 3
```
job.repeatAt('30 10 * * *').save()
```

Please point me out what wrong am I doing.
Thank you.
